server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.reference.doc.lucene;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.StoredFieldVisitor;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.compress.NotXContentException;
+import org.elasticsearch.common.io.stream.ByteBufferStreamInput;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.jetbrains.annotations.NotNull;
+
+import io.crate.common.collections.Maps;
+import io.crate.execution.dml.ArrayIndexer;
+import io.crate.metadata.ColumnIdent;
+import io.crate.metadata.Reference;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.metadata.doc.SysColumns;
+import io.crate.types.ArrayType;
+import io.crate.types.DataType;
+import io.crate.types.GeoShapeType;
+import io.crate.types.ObjectType;
+
+/**
+ * Loads column values from stored fields and presents them as a java Map
+ */
+public class ColumnFieldVisitor extends StoredFieldVisitor {
+
+    private final Map<String, Field> fields = new HashMap<>();
+    private final Set<ColumnIdent> droppedColumns;
+    private final SourceParser storedSourceParser;
+
+    /**
+     * Creates a new ColumnFieldVisitor for the given table
+     */
+    public ColumnFieldVisitor(DocTableInfo table) {
+        this.droppedColumns
+            = table.droppedColumns().stream().map(Reference::column).collect(Collectors.toUnmodifiableSet());
```

Comment:
Looking at the usages of `droppedColumns` I think we should consider changing the representation inside `DocTableInfo` to `Map<ColumnIdent, Reference>`. It looks like the `Set<Reference>` representation is almost never a fit.

Reply:
+1, I'll open a separate PR to handle this.

Reply:
Changing this to a Map doesn't really help things, unfortunately.  Generally speaking, callers either want to stream the References or convert them to something that can be accessed via the ColumnIdent, so everything either gets more complex (`droppedColumns()` becomes `droppedColumns.values()`) or stays about the same.  `ColumnFieldVisitor` is the only caller that becomes simpler, so I don't think it's worth it.